### DISCUSSION TITLE
Use basic auth protected API as tile source / pydeck

### DISCUSSION THREAD

**Discussion by battamatt** [visgl/deck.gl#6223] [Q&A] [answered]
Hi everyone,

I would like to render a TerrainLayer in Jupyter using pydeck requesting data from a proprietary tile server URL that is protected by a username and password and basic auth. Placing the username and the password in the URL does not seem to work. Is there a way to do this with the load_options parameter and the expression parser or any other workaround?

Thanks,
Mattia

**Comment by kylebarron** [comment]
I believe pydeck passes any extra keyword arguments to the `pydeck.Layer` class on to the JS layer instance. So it might work to do

```py
Layer(
	'TerrainLayer',
	...,
	'loadOptions'={
	   'fetch': {
	        'headers': {
	          'Authorization': 'Authorization string',
	        }
	      }
	    }

```

**Reply by battamatt** [reply]
Thanks for your reply! Unfortunately, 
this particular syntax did not work for me. The request is registered by the server, but as far as I can see without any Authorization entry.

**Reply by kylebarron** [reply]
> The request is registered by the server

I can't tell what that means. You should be able to look at HTTP requests from your browser to see if the header value is included in the requests.

**Reply by battamatt** [reply]
Thanks, I was able to change my server settings which finally made your suggestion work.

**Comment by ibgreen** [comment]
load_options should have a `fetch` field that takes an object with fetch options that are passed to the underlying fetch call. It can be used to set headers etc.

Haven't tested via pydeck but in principle it should work

https://developer.mozilla.org/en-US/docs/Web/API/fetch#parameters

**Reply by battamatt** [reply]
Thanks for the hint! Could you maybe provide an example python code that shows how to implement this?

**Comment by ajduberstein** [comment, marked as answer]
@DrKarlMarxx Modifying [the example in the docs](https://pydeck.gl/gallery/terrain_layer.html) to use @ibgreen's suggestion–

```python
import pydeck as pdk
import os

# Import Mapbox API Key from environment
MAPBOX_API_KEY = os.environ["MAPBOX_API_KEY"]

# AWS Open Data Terrain Tiles
TERRAIN_IMAGE = "https://s3.amazonaws.com/elevation-tiles-prod/terrarium/{z}/{x}/{y}.png"

# Define how to parse elevation tiles
ELEVATION_DECODER = {"rScaler": 256, "gScaler": 1, "bScaler": 1 / 256, "offset": -32768}

SURFACE_IMAGE = f"https://api.mapbox.com/v4/mapbox.satellite/{{z}}/{{x}}/{{y}}@2x.png?access_token={MAPBOX_API_KEY}"

terrain_layer = pdk.Layer(
    "TerrainLayer",
    elevation_decoder=ELEVATION_DECODER,
    texture=SURFACE_IMAGE,
    elevation_data=TERRAIN_IMAGE,
    # Additional parameter here
    load_options={
        'fetch': {
	    'headers': {
	        'Authorization': 'your_token_here'
            }
	}
    }
)

view_state = pdk.ViewState(latitude=46.24, longitude=-122.18, zoom=11.5, bearing=140, pitch=60)

r = pdk.Deck(terrain_layer, initial_view_state=view_state)

r.to_html("terrain_layer.html")
```

You can open up the "Network" tab of your browser's developer console, click on the request, and see the modified header.

**Reply by battamatt** [reply]
Thanks, this helped me finding the reason why my server did not respond correctly!
